Add support for java11 in gae

SudharakaP opened the pull request:
This migrates our existing GAE generator from Java 8 to Java 11. Note that this will remove all support for Java 8 GAE config file generation (in other words after this fix is merged users will only be able to generate Java 11 GAE configs). I thought this is the best since giving users the ability to choose between Java 8 and Java 11 seems messy since we have migrated other parts of jhipster-generator (such as the travis build to java 11) completely. Also the configs and the way GAE deployments work is very different from Java 8 to Java 11; and Java 11 uses JAR deployment which simplifies matters a lot.

Fix #10196

-   Please make sure the below checklist is followed for Pull Requests.

-   [x] [Travis tests](https://travis-ci.org/jhipster/generator-jhipster/pull_requests) are green
-   [x] Tests are added where necessary
-   [x] Documentation is added/updated where necessary
-   [x] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/master/CONTRIBUTING.md) are followed

<!--
Please also reference the issue number in a commit message to [automatically close the related Github issue](https://help.github.com/articles/closing-issues-via-commit-messages/)

Note: It is also possible to add `[skip ci]` to your commit message to skip Travis tests
-->

PierreBesson commented on generators/gae/index.js:
```diff
@@ -363,17 +366,41 @@ module.exports = class extends BaseGenerator {
-
                 exec(
-                    `gcloud sql instances list  --format='value[separator=":"](project,region,name)' --project="${this.gcpProjectId}"`,
+                    `gcloud sql instances list  --format="value[separator=":"](project,region,name)" --project="${this.gcpProjectId}"`,
```
I'm pretty sure there is a mistake here with the `"`

SudharakaP replied:
Thanks. Corrected 😄

PierreBesson commented on generators/gae/templates/app.yaml.ejs:
```diff
@@ -0,0 +1,20 @@
+    max_instances: <%= gaeMaxInstances %>
+    <%_ } _%>
+<%_ } else if (gaeScalingType === 'automatic') { _%>
+automatic_scaling:
```
I have an error when answering the default values in the prompt of `min_instance: 0` and `max_instance: 0`, in this cas the `automatic_scaling:` block should disappear entirely. https://cloud.google.com/appengine/docs/flexible/java/reference/app-yaml#automatic_scaling

SudharakaP replied:
I've made changes to this. I think one should be allowed to choose max_instances: 0 and/or min instances: 0. These are allowed settings in GCP and I don't think we need the block to disappear. If max_instances: 0 is chosen the setting will be disabled anyways. I've made an error in my previous commit such that when 0 is chosen the max/min_instances blocks doesn't appear. 

Let me know your thoughts; I also see that the reference doc you pointed seems to the flexible environment not the standard one which is what we support; the correct doc is; https://cloud.google.com/appengine/docs/standard/java11/config/appref#automatic_scaling
The way standard handles max/min_instances is a bit different. 

Supporting flexible is a different discussion which I'll bring up in perhaps another PR. 😄

PierreBesson replied:
Yes you are correct on both points.

PierreBesson commented on generators/gae/templates/app.yaml.ejs:
```diff
@@ -0,0 +1,20 @@
+runtime: java11
```
Maybe start this file with the following:
```
# ========================================================================
# App Engine Configuration.
# Full reference is available at:
# https://cloud.google.com/appengine/docs/flexible/java/reference/app-yaml
# ========================================================================

SudharakaP replied:
Thanks. Done 😄

PierreBesson commented:
@SudharakaP For some reason when running `jhipster gae`, I have an error when generating `application-prod-gae.yml`

```
SyntaxError: Nested mappings are not allowed in compact mappings (18:19)
  16 |         url: jdbc:mysql://google/gaedemo?cloudSqlInstance=proven-serenity-238515:northamerica-northeast1:demo1
  17 | &socketFactory=com.google.cloud.sql.mysql.SocketFactory&useSSL=false
> 18 |         username: root
     |                   ^^^^^^^^^^^^^
> 19 |         password: 
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> 20 |         hikari:
     | ^
  21 |             maximumPoolSize: 8
  22 | 
    at e (/home/pierre/github/generator-jhipster/node_modules/prettier/parser-yaml.js:1:261)
    at Object.parse (/home/pierre/github/generator-jhipster/node_modules/prettier/parser-yaml.js:1:135253)
    at Object.parse$2 [as parse] (/home/pierre/github/generator-jhipster/node_modules/prettier/index.js:10639:19)
    at coreFormat (/home/pierre/github/generator-jhipster/node_modules/prettier/index.js:13856:23)
    at format (/home/pierre/github/generator-jhipster/node_modules/prettier/index.js:14115:73)
    at formatWithCursor (/home/pierre/github/generator-jhipster/node_modules/prettier/index.js:14131:12)
    at /home/pierre/github/generator-jhipster/node_modules/prettier/index.js:42399:15
    at Object.format (/home/pierre/github/generator-jhipster/node_modules/prettier/index.js:42418:12)
    at prettier.resolveConfig.then.options (/home/pierre/github/generator-jhipster/generators/generator-transforms.js:44:39)
```
Can you add quotes in the application-prod-gae.yml template aound values like this:
```
spring:
    datasource:
        type: com.zaxxer.hikari.HikariDataSource
        url: "jdbc:mysql://google/<%=gcpCloudSqlDatabaseName%>?cloudSqlInstance=<%=gcpCloudSqlInstanceName%>&socketFactory=com.google.cloud.sql.mysql.SocketFactory&useSSL=false"
        username: <%=gcpCloudSqlUserName%>
        password: <%=gcpCloudSqlPassword%>
        hikari:
            maximumPoolSize: 8
    datasource:
        type: com.zaxxer.hikari.HikariDataSource
        url: "jdbc:postgresql://google/<%=gcpCloudSqlDatabaseName%>?cloudSqlInstance=<%=gcpCloudSqlInstanceName%>&socketFactory=com.google.cloud.sql.postgres.SocketFactory"
        username: "<%=gcpCloudSqlUserName%>"
        password: "<%=gcpCloudSqlPassword%>"
```

saturnism commented:
@SudharakaP this is awesome! thank you.

I didn't see app engine plugin version being updated. It may need bump. https://cloud.google.com/appengine/docs/standard/java11/using-maven#adding_the_maven_and_eclipse_jetty_maven_plugins

/cc @ludoch

SudharakaP commented:
@saturnism : You are welcome. I'll check and fix all the things mentioned by Pierre soon. Then hopefully we'll have a working app engine generator. Although I do have some more plans for this one. Will create some more prs in the future. :smile:

SudharakaP commented:
> @SudharakaP For some reason when running `jhipster gae`, I have an error when generating `application-prod-gae.yml`
> 
> ```
> SyntaxError: Nested mappings are not allowed in compact mappings (18:19)
>   16 |         url: jdbc:mysql://google/gaedemo?cloudSqlInstance=proven-serenity-238515:northamerica-northeast1:demo1
>   17 | &socketFactory=com.google.cloud.sql.mysql.SocketFactory&useSSL=false
> > 18 |         username: root
>      |                   ^^^^^^^^^^^^^
> > 19 |         password: 
>      | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> > 20 |         hikari:
>      | ^
>   21 |             maximumPoolSize: 8
>   22 | 
>     at e (/home/pierre/github/generator-jhipster/node_modules/prettier/parser-yaml.js:1:261)
>     at Object.parse (/home/pierre/github/generator-jhipster/node_modules/prettier/parser-yaml.js:1:135253)
>     at Object.parse$2 [as parse] (/home/pierre/github/generator-jhipster/node_modules/prettier/index.js:10639:19)
>     at coreFormat (/home/pierre/github/generator-jhipster/node_modules/prettier/index.js:13856:23)
>     at format (/home/pierre/github/generator-jhipster/node_modules/prettier/index.js:14115:73)
>     at formatWithCursor (/home/pierre/github/generator-jhipster/node_modules/prettier/index.js:14131:12)
>     at /home/pierre/github/generator-jhipster/node_modules/prettier/index.js:42399:15
>     at Object.format (/home/pierre/github/generator-jhipster/node_modules/prettier/index.js:42418:12)
>     at prettier.resolveConfig.then.options (/home/pierre/github/generator-jhipster/generators/generator-transforms.js:44:39)
> ```
> 
> Can you add quotes in the application-prod-gae.yml template aound values like this:
> 
> ```
> spring:
>     datasource:
>         type: com.zaxxer.hikari.HikariDataSource
>         url: "jdbc:mysql://google/<%=gcpCloudSqlDatabaseName%>?cloudSqlInstance=<%=gcpCloudSqlInstanceName%>&socketFactory=com.google.cloud.sql.mysql.SocketFactory&useSSL=false"
>         username: <%=gcpCloudSqlUserName%>
>         password: <%=gcpCloudSqlPassword%>
>         hikari:
>             maximumPoolSize: 8
>     datasource:
>         type: com.zaxxer.hikari.HikariDataSource
>         url: "jdbc:postgresql://google/<%=gcpCloudSqlDatabaseName%>?cloudSqlInstance=<%=gcpCloudSqlInstanceName%>&socketFactory=com.google.cloud.sql.postgres.SocketFactory"
>         username: "<%=gcpCloudSqlUserName%>"
>         password: "<%=gcpCloudSqlPassword%>"
> ```

@PierreBesson : Thanks for all the code reviews; I've fixed them all in my latest commit. Please let me know if you see any further issues. 😄 

> @SudharakaP this is awesome! thank you.
> 
> I didn't see app engine plugin version being updated. It may need bump. https://cloud.google.com/appengine/docs/standard/java11/using-maven#adding_the_maven_and_eclipse_jetty_maven_plugins
> 
> /cc @ludoch

@saturnism : Thanks for pointing this out. I've updated the plugin version as well to 2.1.0. 😄

PierreBesson commented:
Thank you @SudharakaP for the hard work. Very appreciated. Looking forward to review your next contributions on app engine support. BTW you can remove the `BETA` log at startup as it doesn't do justice to this subgen.

SudharakaP commented:
@PierreBesson : You are welcome; it's fun contributing to jhipster. 

Btw, I would suggest keeping the ```BETA``` log for a week or so until I get some other corner cases right. I think there's still issues which could cause confusion in this sub-generator. I've created a new issue for discussion; https://github.com/jhipster/generator-jhipster/issues/10331. 

cc: @saturnism 

Please feel free to add any suggestions. :smile:

jdubois commented:
@SudharakaP thanks a lot for all your work and contributions!!! I also saw you sponsored the project ( https://opencollective.com/sudharaka-palamakumbura )!! Do you have a Twitter handle, so I can thank you publicly?

SudharakaP commented:
@jdubois: You are welcome, and hope it makes a difference. I don't use Twitter :smile:
